modules/core/src/main/java/org/apache/ignite/internal/util/subscription/IterableToPublisherAdapter.java

```diff
@@ -0,0 +1,193 @@
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.util.subscription;
+
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodHandles.Lookup;
+import java.lang.invoke.VarHandle;
+import java.util.Iterator;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Flow.Publisher;
+import java.util.concurrent.Flow.Subscriber;
+import java.util.concurrent.Flow.Subscription;
+
+/**
+ * An adapter that issues a new iterator for every subscription and drains that iterator on a given
+ * executor, emitting up to given {@code batchSize} entries at time. The drain task will be re-scheduled
+ * until demand is fulfilled or iterator has no more items.
+ *
+ * @param <T> The type of the entry this publisher will emit.
+ */
+public class IterableToPublisherAdapter<T> implements Publisher<T> {
+    private final Iterable<T> iterable;
+    private final Executor executor;
+    private final int batchSize;
+
+    /**
+     * Constructor.
+     *
+     * @param iterable An iterable to issue iterator for every incoming subscription.
+     * @param executor This executor will be used to drain iterator and supply entries to the subscription.
+     * @param batchSize An amount of entries to supply during a single iteration. It's always good idea
+     *      to provide some reasonable value here in order to give am ability to other publishers which share the same
+     *      executor to make progress.
+     */
+    public IterableToPublisherAdapter(Iterable<T> iterable, Executor executor, int batchSize) {
+        this.iterable = iterable;
+        this.executor = executor;
+        this.batchSize = batchSize;
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super T> subscriber) {
+        Iterator<T> it = iterable.iterator();
+
+        Subscription subscription = new SubscriptionImpl<>(it, subscriber, executor, batchSize);
+
+        subscriber.onSubscribe(subscription);
+    }
+
+    @SuppressWarnings("FieldMayBeFinal")
+    private static class SubscriptionImpl<T> implements Subscription {
+        private static final VarHandle CANCELLED_HANDLE;
+        private static final VarHandle REQUESTED_HANDLE;
+        private static final VarHandle WIP_HANDLE;
```

Comment:
Personally I don't see the point of using varhandles instead of atomics here.
From my point of view the code with atomics will look shorter and cleaner.

Reply:
well, it's part of the core library, thus may be used on a hot path. It's better to avoid creation of unnecessary objects